src/Shared/PropertyAsParameterInfo.cs

```diff
@@ -182,7 +183,20 @@ public override bool IsDefined(Type attributeType, bool inherit)
             _underlyingProperty.IsDefined(attributeType, inherit);
     }
 
-    public new bool IsOptional => HasDefaultValue || NullabilityInfo.ReadState != NullabilityState.NotNull;
+    public new bool IsOptional => NullabilityInfo.ReadState switch
+    {
+        // Anything nullable is optional
+        NullabilityState.Nullable => true,
+        // In an oblivious context, the required modifier makes
+        // members non-optional
+        NullabilityState.Unknown => !GetCustomAttributes(true).OfType<RequiredMemberAttribute>().Any(),
```

Comment:
`GetCustomAttributes(true)` will try merge attributes from both constructor argument and property. maybe you can change it to `_underlyingProperty.GetCustomAttributes(true)` since `RequiredMemberAttribute` is not allowed in parameters.